core/src/main/scala/akka/kafka/internal/SingleSourceLogic.scala

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.kafka.internal
+
+import akka.kafka.Subscription.{Assignment, AssignmentWithOffset, TopicSubscription, TopicSubscriptionPattern}
+import akka.kafka.scaladsl.KafkaAsyncConsumer
+import akka.kafka.{ConsumerSettings, Subscription}
+import akka.stream.stage.{GraphStageLogic, OutHandler}
+import akka.stream.{ActorMaterializer, SourceShape}
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.common.TopicPartition
+
+import scala.annotation.tailrec
+import scala.util.{Failure, Success, Try}
+
+private[kafka] abstract class SingleSourceLogic[K, V, Msg](
+    shape: SourceShape[Msg],
+    settings: ConsumerSettings[K, V],
+    subscription: Subscription
+) extends GraphStageLogic(shape) with PromiseControl with MessageBuilder[K, V, Msg] {
+  var consumer: KafkaAsyncConsumer[K, V] = _
+  var tps = Set.empty[TopicPartition]
+  var buffer: Iterator[ConsumerRecord[K, V]] = Iterator.empty
+  var requested = false
+  implicit lazy val ec = ActorMaterializer.downcast(materializer).executionContext
+
+  override def preStart(): Unit = {
+    super.preStart()
+    setKeepGoing(true)
+    consumer = KafkaAsyncConsumer[K, V](settings)(ActorMaterializer.downcast(materializer).system)
```

Comment:
You can call `getStageActor` which takes a "receive" block, which is part of your stage, and returns an ActorRef that will point to that receive block and which you can share with the external world, posing like you were an actor.